client.go

```diff
@@ -253,12 +180,13 @@ func (c *Client) exchange(ctx context.Context, m *Msg, a string) (r *Msg, rtt ti
 	}
 
 	co.TsigSecret = c.TsigSecret
-	co.SetWriteDeadline(deadlineOrTimeoutOrCtx(ctx, deadline, c.writeTimeout()))
+	// write with the appropriate write timeout
+	co.SetWriteDeadline(time.Now().Add(c.getTimeoutForRequest(dialer, c.writeTimeout())))
```

Comment:
why are you removing/rewriting deadlineOrTimeotOrCtx? Because we don't have ctx anymore?

Reply:
Yes but not only. Background: this part of the diff went through a couple of iterations, where I first removed the function, and eventually reinstated it. When I reinstated it, the content was slightly different, and I was looking for a name that was more descriptive of its behaviour.

Reply:
Can we *not* change it now, so we can just focus on getting a net.Dialer in ?

Reply:
ok, will revert this part

Reply:
Went through the code again, and refreshed my memory. I remembered other reasons to rewrite that function:

* all the parameters are now timeouts, not deadlines (since context is gone, and Dialer has no deadline either). This makes them easily comparable, which is not the case when mixing deadlines and timeouts
* context is gone, deadline is gone, dialer is coming, so basically this function would have a different name and a different content too. Not sure if it still makes sense to leave it as it is, as most of its content has to change anyway

As per the above I still believe that this function should be rewritten within this PR. Since we are moving a Dialer under Client, it could be done like this:

* remove deadlineOrTimeoutOrCtx anyway (it's not part of the API, it's unexported, and as said above it's not relevant anymore in the original form)
* leave this new method (whatever name we want to give it) attached to Client, and simplify it by just passing a time.Duration. The dialer timeout can be accessed inside as `c.Dialer.Timeout`. In other words `func (c *Client) getTimeoutForRequest(timeout time.Duration) time.Duration`.

What do you think of the above?

Reply:
done as I suggested above so you can review the new diff, then decide how to proceed. Diff incoming